### PR TITLE
Add cond-dt, above, below, between, outside streams

### PR DESCRIPTION
**Pull request by mcorbin**
cf #788 

This PR adds a few new streams. 

## above example

"Takes a number `threshold` and a time period in seconds `dt`.
If the condition `(> (:metric event) threshold)` is valid for all events received during at least the period `dt`, valid events received after the `dt` period will be passed on until an invalid event arrives.
`:metric` should not be nil (it will produce exceptions)."

Let's begin with an example (a test case) :

```clojure
(test-stream (above 10 5) [{:time 0
                            :metric 11}
                           {:time 7
                            :metric 4}
                           {:time 9
                            :metric 11}
                           {:time 15
                            :metric 12}] [{:time 15
                                           :metric 12}])
```
In this test, the threshold is 10, dt is 5.
- At time 0, :metric is 11 (so > 10, event is valid)
- At time 7, :metric is 4 (< 10, invalid)
- At time 9, :metric is 11 (valid)
- At time 15, :metric is 12 (valid). During 6 seconds (15 - 9), :metric was > to threshold => above passes on this event.

Each new valid event will be forwarded to children, until an invalid event arrives.

## below, between, outside

below check if :metric is < to a threshold during dt seconds, between if :metric is inside ]low, high[ during dt seconds, outside if :metric is outside [low, high] during dt seconds.

## cond-dt

All these streams used cond-dt under the hood. cond-dt takes a function of events, a duration, and children. For example, the above stream is just :

```clojure
(apply cond-dt (fn [event] (> (:metric event) threshold)) dt children)
```
cond-dt can be used to creates new streams by users.

## limitations

Like in all Riemann streams : events arrives from the past. You can check [this test case](https://github.com/riemann/riemann/pull/789/files#diff-f4d4815a1ac7dab86942b0dda98a1d7cR1764) for example.

What do you think about these streams ? Do you think they have their place in Riemann ?

### CODE REVIEW

File: src/riemann/streams.clj [hunk 1]
```diff
@@ -2089,3 +2089,67 @@
...
+                          :default state)))
+          (when (and valid-event
+                      ;; we already had an ok event
+                     ok
+                     ;; check is current time > first ok event + dt
+                     (> event-time (+ changed-state-time dt)))
+            (call-rescue event children)))))))
+
+(defn above
+  "Takes a number `threshold` and a time periode in seconds `dt`. Detects if the `:metric` value is > to thresold during dt seconds using cond-dt.
```
Comment (boernd) [inline src/riemann/streams.clj]: Typos (periode and thresold)

Comment (mcorbin) [inline src/riemann/streams.clj]: fixed, thanks

File: src/riemann/streams.clj [hunk 2]
```diff
@@ -2089,3 +2089,66 @@
                             (dissoc ::clock-skew-timestamp)
                             (assoc :metric delta))]
                 (call-rescue event children)))))))
+
+(defn cond-dt
+  "Takes a function of events `f` and a time period `dt` in seconds. Detects if `(f event)` is true during dt seconds.
```
Comment (Mogztter) [inline src/riemann/streams.clj]: I think the documentation is a bit misleading.
What about:

> A stream which detects if a condition `(f event)` is true during `dt` seconds.
Takes a function of events `f` and a time period `dt` in seconds.

> If the condition is valid for all events received during at least the period `dt`, valid events received after the `dt` period will be passed on until an invalid event arrives.
Skips too old events or events without time.

Disclaimer: I'm not a native English speaker :wink:

Comment (mcorbin) [inline src/riemann/streams.clj]: I like your description, i updated the PR ;) Thanks !

File: src/riemann/streams.clj [hunk 3]
```diff
@@ -2089,3 +2089,68 @@
                             (dissoc ::clock-skew-timestamp)
                             (assoc :metric delta))]
                 (call-rescue event children)))))))
+
+(defn cond-dt
+  "A stream which detects if a condition `(f event)` is true during `dt` seconds.
+  Takes a function of events `f` and a time period `dt` in seconds.
+&
+    If the condition is valid for all events received during at least the period `dt`, valid events received after the `dt` period will be passed on until an invalid event arrives.
+    Skips too old events or events without time."
```
Comment (jamtur01) [inline src/riemann/streams.clj]: `Skips events that are too old or that do not have a time stamp.`

File: src/riemann/streams.clj [hunk 4]
```diff
@@ -2089,3 +2089,68 @@
...
+          (when (and valid-event
+                      ;; we already had an ok event
+                     ok
+                     ;; check is current time > first ok event + dt
+                     (> event-time (+ changed-state-time dt)))
+            (call-rescue event children)))))))
+
+(defn above
+  "Takes a number `threshold` and a time period in seconds `dt`. Detects if the `:metric` value is > to threshold during dt seconds using cond-dt.
+  :metric should not be nil."
```
Comment (jamtur01) [inline src/riemann/streams.clj]: What happens if the `:metric` is nil? Perhaps explain that?

File: src/riemann/streams.clj [hunk 5]
```diff
@@ -2089,3 +2089,68 @@
...
+                          :default state)))
+          (when (and valid-event
+                      ;; we already had an ok event
+                     ok
+                     ;; check is current time > first ok event + dt
+                     (> event-time (+ changed-state-time dt)))
+            (call-rescue event children)))))))
+
+(defn above
+  "Takes a number `threshold` and a time period in seconds `dt`. Detects if the `:metric` value is > to threshold during dt seconds using cond-dt.
```
Comment (jamtur01) [inline src/riemann/streams.clj]: `the `:metric` value is > to threshold during dt seconds using cond-dt` - I don't understand what you're trying to say here?

Also inconsistent use of back tick quoting.

File: src/riemann/streams.clj [hunk 6]
```diff
@@ -2089,3 +2089,68 @@
...
+  :metric should not be nil."
+  [threshold dt & children]
+  (apply cond-dt (fn [event] (> (:metric event) threshold)) dt children))
+
+(defn below
+  "Takes a number `threshold` and a time period in seconds `dt`. Detects if the `:metric` value is < to threshold during dt seconds using cond-dt.
```
Comment (jamtur01) [inline src/riemann/streams.clj]: As above.

File: src/riemann/streams.clj [hunk 7]
```diff
@@ -2089,3 +2089,68 @@
...
+  :metric should not be nil."
+  [threshold dt & children]
+  (apply cond-dt (fn [event] (< (:metric event) threshold)) dt children))
+
+(defn between
+  "Takes two numbers `low` and `high` and a time period in seconds `dt`. Detects if the `:metric` value is > to low AND < to high during dt seconds using cond-dt.
```
Comment (jamtur01) [inline src/riemann/streams.clj]: "Takes two numbers, `low` and `high`, and a time period in seconds, `dt`."

Comment (jamtur01) [inline src/riemann/streams.clj]: Also inconsistent quoting.

Comment (jamtur01) [inline src/riemann/streams.clj]: "Detects if the `:metric` value is > to low AND < to high during dt seconds using cond-dt." - again unclear....

File: src/riemann/streams.clj [hunk 8]
```diff
@@ -2089,3 +2089,68 @@
...
+  :metric should not be nil."
+  [low high dt & children]
+  (apply cond-dt (fn [event] (and (> (:metric event) low)
+                                  (< (:metric event) high))) dt children))
+
+(defn outside
```
Comment (jamtur01) [inline src/riemann/streams.clj]: As above.

File: src/riemann/streams.clj (hunk 3 above)
Comment (mcorbin) [inline src/riemann/streams.clj]: Fixed, thanks !

File: src/riemann/streams.clj (hunk 5 above)
Comment (mcorbin) [inline src/riemann/streams.clj]: I rewrote the description. Sorry for my bad english :) Thanks !

File: src/riemann/streams.clj (hunk 4 above)
Comment (mcorbin) [inline src/riemann/streams.clj]: I explained it, thanks !

File: src/riemann/streams.clj (hunk 6 above)
Comment (mcorbin) [inline src/riemann/streams.clj]: I modified the description, thanks !

File: src/riemann/streams.clj (hunk 7 above)
Comment (mcorbin) [inline src/riemann/streams.clj]: Fixed, thanks !

Comment (mcorbin) [inline src/riemann/streams.clj]: Fixed, thanks !

Comment (mcorbin) [inline src/riemann/streams.clj]: I modified the description, thanks !

File: src/riemann/streams.clj (hunk 8 above)
Comment (mcorbin) [inline src/riemann/streams.clj]: Fixed, thanks !

**Comment by mcorbin** [comment]
@aphyr @pyr What do you think about it ? I think the use case "is (f event) true during dt seconds" is very common. I tried to do a clean implementation of the functionality, but maybe we can do better ?

**Comment by aphyr** [comment]
So... I don't doubt that this is useful, but as far as Riemann's core goes, I try to keep functions limited to those that will have broad applicability with obvious, composable semantics. These seem like... specific sorts of functions that have lots of variants depending on use case, and I think the best thing to do there is to let people implement them themselves via (fixed-time-window (smap (fn [events] (every? pred event))) or similar.

**Comment by mcorbin** [comment]
The problem i have with `(fixed-time-window (smap (fn [events] (every? pred event)))` is this:

|----TW1----|----TW2---|              
|ooooXXXX|XXXXoooo|
|-----------------------------> t

I will miss the problem because it happens in 2 time windows. 

But i can also release it as a plugin ;)

**Comment by aphyr** [comment]
Or moving-time-window, or sreduce directly.

**Comment by mcorbin** [comment]
Ok, i close it (and will move the code in a plugin). Thanks you ;)

**Pull request closed without merging**